### PR TITLE
[v2] [cache-dir] Fix navigation after reload

### LINKED ISSUE #7261: [v2] Refreshing a page then hitting back (browser back button) leads to 404
## Description

I have a local project that is using the v2 of Gatsby. Eevrything works great except for the fact that a page refresh and a back button on the browser always leads to a 404. On that 404 page Gatsby displays a list of routes and the actual route (the one that responded with 404) is in that list.

```
http://localhost:8000/categories/release/
-----
Gatsby.js development 404 page
There's not a page yet at /categories/release/

Create a React.js component in your site directory at src/pages/categories/release.js and this page will automatically refresh to show the new page component you created.

If you were trying to reach another page, perhaps you can find it below.

Pages (7)
/tags/cards/
/tags/example/
/tags/release/
/categories/release/
/404/
/
/404.html
```

### Environment

```
gatsby info --clipboard

  System:
    OS: Linux 4.17 Fedora 28 (Workstation Edition) 28 (Workstation Edition)
    CPU: x64 Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
    Shell: 4.4.23 - /bin/bash
  Binaries:
    Node: 10.7.0 - /usr/bin/node
    Yarn: 1.9.4 - /usr/bin/yarn
    npm: 6.1.0 - /usr/bin/npm
  Browsers:
    Chrome: 68.0.3440.75
    Firefox: 61.0.1
  npmPackages:
    gatsby: next => 2.0.0-beta.97 
    gatsby-1-config-css-modules: next => 1.0.10-13 
    gatsby-image: next => 2.0.0-beta.7 
    gatsby-plugin-google-fonts: 0.0.4 => 0.0.4 
    gatsby-plugin-mailchimp: 2.2.3 => 2.2.3 
    gatsby-plugin-netlify: ^2.0.0-beta.6 => 2.0.0-beta.6 
    gatsby-plugin-react-helmet: next => 3.0.0-beta.4 
    gatsby-plugin-sass: next => 2.0.0-beta.10 
    gatsby-plugin-sharp: next => 2.0.0-beta.7 
    gatsby-plugin-styled-components: next => 3.0.0-beta.3 
    gatsby-plugin-typescript: next => 2.0.0-beta.9 
    gatsby-source-contentful: next => 2.0.1-beta.15 
    gatsby-source-filesystem: next => 2.0.1-beta.10 
    gatsby-transformer-sharp: next => 2.1.1-beta.6 
  npmGlobalPackages:
    gatsby-cli: 1.1.58
    gatsby: 2.0.0-beta.54

```

### File contents (if changed)

`gatsby-config.js`: 

```
let contentfulConfig, mailchimpConfig;

try {
	contentfulConfig = require('./config/contentful.json');
} catch (_) {
	contentfulConfig = {
		spaceId: process.env.CONTENTFUL_SPACE_ID,
		accessToken: process.env.CONTENTFUL_DELIVERY_TOKEN,
	}
}

try {
	mailchimpConfig = require('./config/mailchimp.json')
} catch (_) {
	mailchimpConfig = {
		endpoint: process.env.MAILCHIMP_ENDPOINT
	}
}

module.exports = {
	plugins: [
		'gatsby-plugin-react-helmet',
		`gatsby-plugin-typescript`,
		`gatsby-transformer-sharp`,
		`gatsby-plugin-sharp`,
		`gatsby-plugin-sass`,
		`gatsby-plugin-styled-components`,
		{
			resolve: `gatsby-plugin-google-fonts`,
			options: {
				fonts: [
					`Roboto:300,300i,400,400i,500,500i,600,600i`,
				],
			},
		},
		{
			resolve: `gatsby-source-contentful`,
			options: contentfulConfig,
		},
		{
			resolve: 'gatsby-plugin-mailchimp',
			options: mailchimpConfig,
		},
		`gatsby-plugin-netlify`, // make sure to keep it last in the array
	],
};

```

`package.json`: 

```
{
  "name": "",
  "description": "",
  "version": "1.0.0",
  "author": "",
  "dependencies": {
    "@blueprintjs/core": "^3.0.1",
    "autoprefixer": "^9.0.1",
    "babel-plugin-styled-components": "^1.5.1",
    "extract-text-webpack-plugin": "^3.0.2",
    "gatsby": "next",
    "gatsby-1-config-css-modules": "next",
    "gatsby-image": "next",
    "gatsby-plugin-google-fonts": "0.0.4",
    "gatsby-plugin-mailchimp": "2.2.3",
    "gatsby-plugin-netlify": "^2.0.0-beta.6",
    "gatsby-plugin-react-helmet": "next",
    "gatsby-plugin-sass": "next",
    "gatsby-plugin-sharp": "next",
    "gatsby-plugin-styled-components": "next",
    "gatsby-plugin-typescript": "next",
    "gatsby-source-contentful": "next",
    "gatsby-source-filesystem": "next",
    "gatsby-transformer-sharp": "next",
    "grid-styled": "^5.0.2",
    "install": "^0.12.1",
    "lodash": "^4.17.10",
    "lodash.debounce": "^4.0.8",
    "moment": "^2.22.2",
    "node-sass": "^4.9.2",
    "npm": "^6.3.0",
    "prismjs": "^1.15.0",
    "react": "^16.4.1",
    "react-accessible-accordion": "^2.4.4",
    "react-custom-scrollbars": "^4.2.1",
    "react-dom": "^16.4.1",
    "react-helmet": "^5.2.0",
    "react-player": "^1.6.4",
    "react-prism": "^4.3.2",
    "react-redux": "^5.0.7",
    "react-share": "^2.2.0",
    "react-slick": "^0.23.1",
    "remark": "^9.0.0",
    "remark-react": "^4.0.3",
    "slick-carousel": "^1.8.1",
    "styled-components": "^3.3.3",
    "tinycolor2": "^1.4.1"
  },
  "keywords": [
    "static site",
    "cards",
    "smart cards",
    "blog"
  ],
  "license": "MIT",
  "scripts": {
    "build": "gatsby build",
    "develop": "gatsby develop",
    "format": "prettier --write '**/*.js'",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "devDependencies": {
    "@types/grid-styled": "^4.2.0",
    "@types/lodash": "^4.14.115",
    "@types/moment": "^2.13.0",
    "@types/node": "^10.5.5",
    "@types/prismjs": "^1.9.0",
    "@types/prop-types": "^15.5.4",
    "@types/reach__router": "^1.0.0",
    "@types/react": "^16.4.7",
    "@types/react-dom": "^16.0.6",
    "@types/react-helmet": "^5.0.6",
    "@types/react-redux": "^6.0.5",
    "@types/react-share": "^2.1.1",
    "@types/react-transition-group": "^2.0.13",
    "@types/tinycolor2": "^1.4.1",
    "prettier": "^1.13.7",
    "redux-devtools-extension": "^2.13.5"
  },
  "repository": {
    "type": "git",
    "url": ""
  }
}

```

`gatsby-node.js`: 

```
const _ = require('lodash');
const Promise = require('bluebird');
const path = require('path');
const {createFilePath} = require(`gatsby-source-filesystem`);
const {createRemoteFileNode} = require(`gatsby-source-filesystem`);

const getTags = (edges) => {
	const tags = [];

	_.each(edges, edge => {
		if (_.get(edge, "node.tags")) {
			for (let i = 0; i < edge.node.tags.length; i++) {
				tags.push(edge.node.tags[i]);
			}
		}
	});

	return tags;
};

const getCategories = (edges) => {
	const categories = [];

	_.each(edges, edge => {
		if (_.get(edge, "node.categories")) {
			for (let i = 0; i < edge.node.categories.length; i++) {
				categories.push(edge.node.categories[i]);
			}
		}
	});

	return categories;
};

exports.createPages = ({graphql, actions}) => {
	const {createPage} = actions;

	return new Promise((resolve, reject) => {
		const postTemplate = path.resolve('./src/templates/post/index.tsx');
		const tagTemplate = path.resolve("./src/templates/tag/index.tsx");
		const categoryTemplate = path.resolve("./src/templates/category/index.tsx");

		resolve(graphql(`
				{
					allContentfulPost(sort: {fields: [createdAt], order: DESC}) {
						edges {
							node {
								id
								slug
								
								tags {
									id
									slug
									title
								}

								categories {
									id
									slug
									title
								}

								createdAt
								updatedAt
							}
						}
					}
				}
			`).then(result => {
			if (result.errors) {
				console.error(result.errors);
				reject(result.errors);
			}

			const edges = result.data.allContentfulPost.edges;

			const tags = getTags(edges);
			const categories = getCategories(edges);

			// Create posts and pages.
			_.each(edges, (edge, index) => {
				const slug = edge.node.slug;

				const previous = index === edges.length - 1 ? null : edges[index + 1].node;
				const next = index === 0 ? null : edges[index - 1].node;

				//	create posts pages
				createPage({
					path: `/posts/${slug}/`,
					component: postTemplate,
					context: {
						slug: slug,
						previous,
						next,
					},
				});
			});

			// Make tag pages
			_.uniq(tags).forEach(tag => {
				createPage({
					path: `/tags/${tag.slug}/`,
					component: tagTemplate,
					context: {
						slug: tag.slug
					},
				})
			});

			// Make category pages
			_.uniqBy(categories, 'id').forEach(category => {
				createPage({
					path: `/categories/${category.slug}/`,
					component: categoryTemplate,
					context: {
						slug: category.slug
					},
				})
			});
		}));
	});
};
```

`gatsby-browser.js`: N/A 
`gatsby-ssr.js`: N/A 

NO console errors shown.

### PR DESCRIPTION
**Pull request by davidbailey00**
We added `popstate` listeners to `app.js` and `production-app.js` which load the resources / render the page when back/forward is pressed.

`onPreRouteUpdate` and `onRouteUpdate` aren't currently being called when navigating with back/forward so this will need to be fixed later in order for plugins to work properly.

**Note: wait until #7936 is merged to master before merging this PR, so we can properly test CI beforehand.** ~~Just merged master~~ oops I got mixed up with a different one of DSchau's PRs which was merged

Fixes #7261

### CODE REVIEW

**Comment by KyleAMathews** [comment]
With [@pieh's PR improving our Cypress setup landing](https://github.com/gatsbyjs/gatsby/pull/7816), it'd be nice for this PR to write a test for the reload and back button behavior which _fails_ without the changes made here but then which the changes here do fix .

**Comment by davidbailey00** [comment]
@KyleAMathews I've just added the new test - `waitForRouteChange()` doesn't seem to work with the 404 development page (it resolves after the URL changes but before the 404 appears) so I've gone for the approach of detecting whether the image wrapper element exists, since Cypress will auto-wait for this. I've tested before and after the changes to check it fails and passes correctly.

**Comment by pieh** [comment]
@davidbailey00 `waitForRouteChange` problem is related to problems with navigating through history and firing `onRouteUpdate` - it simply doesn't detect that route has changed in that case and only remember when DOM was updated after reload (before hitting back). Also this currently fail on travis - travis seems much slower, so that's why we added `waitForRouteChange` in the first place, because cypress timeouts were not reliable

**Comment by davidbailey00** [comment]
@pieh In that case perhaps we can override Cypress's default timeout with something like 10 seconds - but I think it's unlikely Travis will be that slow in most cases, given that only two of the tests took longer than 1 second:

image

**Comment by KyleAMathews** [comment]
The plan is to build code & run gatsby-dev for tests but yeah, that makes it more difficult now.

@davidbailey00 could you add tests to here as well? https://github.com/gatsbyjs/gatsby/tree/master/integration-tests/production-runtime

Gatsbygram is running the develop server.

**Comment by pieh** [comment]
Hey, when I was diving deep into `PageRenderer.js` for `onRouteUpdate` fixes for history I noticed we have code there to handle exactly this https://github.com/gatsbyjs/gatsby/blob/0c032cfea2f11886fdb2388d5c05541befa55cfd/packages/gatsby/cache-dir/page-renderer.js#L72 but this isn't correct and we didn't catch this before - this line should not be sync and use `then`:
```diff
-loader.getResourcesForPathnameSync(pathName, pageResources => { 
+loader.getResourcesForPathname(pathName).then(pageResources => {
```
I think changing this and reverting your changes from `production-app.js` and `app.js` should be correct way to fix this. And we still definitely want cypress test you added (which I would later modify a little to use `onRouteUpdate` when I fix that)

**Comment by pieh** [comment]
Change I commented above fixed this for production, but development has a little different setup and we are getting stuck in `JsonStore` component before we even get to `PageRenderer` so I kept development change for now - ultimately I think we could move ad-hoc resources loading from `PageRenderer` somewhere higher in component tree - I'll experiment now for a bit, but won't spend too much time on it

**Comment by davidbailey00** [comment]
@pieh Sounds good, thanks for making the changes! I've got my initial college homework out the way now so I'll try adding the tests requested by Kyle later on.

**Comment by pieh** [comment]
> Just thinking. We should publish our Cypress plugin to npm as people will want to use it when testing their gatsby sites as well. Plus it could lead to more useful Cypress plugins being added.

I agree, I just didn't want to focus on finding proper way of packaging cypress extension + that API is very fresh and I'm not sure if it will stick and if we won't have to change it - I guess it would work if we release it as `0.x` version when nothing is guaranteed ;)

**Comment by pieh** [comment]
note to self: consider moving `ScrollContext` down in component tree as scroll position can reset now before new page have resources (f.e. after reloading page and going back)

File: packages/gatsby/cache-dir/ensure-resources.js [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    const pathname = this.getPathName(location)
+
+    this.state = {
+      location: { ...location },
+      pageResources: loader.getResourcesForPathnameSync(pathname),
+    }
+  }
+
+  static getDerivedStateFromProps({ pageResources, location }, prevState) {
+    if (prevState.location.pathname !== location.pathname) {
```
Comment (pieh) [inline packages/gatsby/cache-dir/ensure-resources.js]: need to make shallow compare - location changes doesnt mean pathname changes

File: packages/gatsby/cache-dir/ensure-resources.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    if (this.state.pageResources.json !== nextState.pageResources.json) {
+      return true
+    }
+
+    // Check if location has changed on a page using internal routing
+    // via matchPath configuration.
+    if (
+      this.props.location.key !== nextProps.location.key &&
```
Comment (pieh) [inline packages/gatsby/cache-dir/ensure-resources.js]: verify `key` is there in `@reach/router`

File: packages/gatsby/cache-dir/ensure-resources.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+      if (url.search) {
+        url.search += `&no-cache=1`
+      } else {
+        url.search = `?no-cache=1`
+      }
+      window.location.replace(url)
+
+      return null
+    } else if (
+      process.env.NODE_ENV !== `production` &&
```
Comment (pieh) [inline packages/gatsby/cache-dir/ensure-resources.js]: verify if this ever happen

**Comment by davidbailey00** [comment]
The Cypress production test I added is failing 80% of the time with Electron 59, but working perfectly every time on the latest Chrome (which we can't use with CI). I spent a while this morning trying to figure out the problem but couldn't find it, and then had to do some other things - I'll have another look now.

**Comment by davidbailey00** [comment]
It works! Travis still fails at the moment because we don't use `gatsby-dev` before testing, but #7936 will fix this (we should merge master to this PR, once #7936 is merged to master, to check it works).

**Comment by KyleAMathews** [comment]
Great work @davidbailey00! Totally fine merging this w/ manual local testing until #7936 is working. I assume @pieh will get this in before I'm awake in the morning but if not, I'll test and merge.

**Comment by pieh** [comment]
We had conversation with @davidbailey00 and there is one more issue that need fixing here - in production sync `loader.getPage` things can fail to get page if we call it before we fetch pages metadata - which is what I think was causing problems on cypress/electron (and not really some bug there). So i currently work on that

**Comment by pieh** [comment]
still TO-DO - move scroll context component to PageRenderer

**Pull request merged**